java.io.IOException is raised when tests use a large class path

**Describe the bug**
All the OS has limits when starting new processes:
- Linux e.g. ARG_MAX (https://www.in-ulm.de/~mascheck/various/argmax/)
- Windows (https://docs.microsoft.com/en-us/troubleshoot/windows-client/shell-experience/command-line-string-limitation#:~:text=The%20maximum%20length%20of%20the,the%20command%20line)

The proposal: generate a class path jar with manifest to work around the limitation


**Expected behavior**
The test must be always executed

**Actual behavior**
When the tests use a large class path an java.io.IOException is thrown (see stack trace below)
org.quickperf.issue.JvmError: java.io.IOException: Cannot run program "C:\Tools\Java\AdoptOpenJDK\Hotspot\JDK_8u272_b10_X64\jre\bin\java": CreateProcess error=206, The filename or extension is too long

Caused by: java.io.IOException: Cannot run program "C:\Tools\Java\AdoptOpenJDK\Hotspot\JDK_8u272_b10_X64\jre\bin\java": CreateProcess error=206, The filename or extension is too long
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at org.quickperf.testlauncher.NewJvmTestLauncher.execute(NewJvmTestLauncher.java:112)
	at org.quickperf.testlauncher.NewJvmTestLauncher.executeTestInNewJvm(NewJvmTestLauncher.java:73)
	at org.quickperf.testlauncher.NewJvmTestLauncher.executeTestMethodInNewJwm(NewJvmTestLauncher.java:43)
	at org.quickperf.junit4.MainJvmAfterJUnitStatement.evaluateBusinessOrTechnicalIssue(MainJvmAfterJUnitStatement.java:77)
	at org.quickperf.junit4.MainJvmAfterJUnitStatement.evaluate(MainJvmAfterJUnitStatement.java:59)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.quickperf.spring.junit4.QuickPerfSpringRunner.runChild(QuickPerfSpringRunner.java:214)
	at org.quickperf.spring.junit4.QuickPerfSpringRunner.runChild(QuickPerfSpringRunner.java:37)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.quickperf.spring.junit4.QuickPerfSpringRunner.run(QuickPerfSpringRunner.java:160)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:68)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:47)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.CommandLineWrapper.main(CommandLineWrapper.java:66)
Caused by: java.io.IOException: CreateProcess error=206, The filename or extension is too long
	at java.lang.ProcessImpl.create(Native Method)
	at java.lang.ProcessImpl.<init>(ProcessImpl.java:444)
	at java.lang.ProcessImpl.start(ProcessImpl.java:139)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 26 more

**To Reproduce**
Define a pom.xml with huge of dependencies and try to run any simple test that  use the class org.quickperf.testlauncher.NewJvmTestLauncher

**Versions**
- QuickPerf: 1.1.0
- JDK: AdoptOpenJDK, Hotspot, JDK_8u272_b10_X64
- OS: Windows
- Database (if SQL annotation bug): not used

**Additional context** (Add any other context about the problem here.)
The implementation is done in my repository clone (checking the contribution guidelines before creating the PR)

--- a/README.md
+++ b/README.md
@@ -407,6 +407,16 @@ Many thanks to all our contributors!
             </a>
             <br/>
             💻
+        </td>        
+        <td align="center">
+            <a href="https://github.com/rsassi">
+                <img src="https://avatars.githubusercontent.com/rsassi" width="100px;" alt="rsassi"/>
+                <br/>
+                <sub><b>rsassi</b></sub>
+            </a>
+            <br/>
+            <a href="https://github.com/quick-perf/quickperf/commits?author=rsassi" title="Bug reports">🐛</a>
+            <a href="https://github.com/quick-perf/quickperf/commits?author=rsassi" title="Code">💻</a>
         </td>
     </tr>
 </table>

--- a/core/src/main/java/org/quickperf/testlauncher/ClassPathUtil.java
+++ b/core/src/main/java/org/quickperf/testlauncher/ClassPathUtil.java
@@ -1,0 +1,101 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Copyright 2019-2022 the original author or authors.
+ */
+package org.quickperf.testlauncher;
+
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.net.MalformedURLException;
+import java.util.jar.Attributes;
+import java.util.jar.JarOutputStream;
+import java.util.jar.Manifest;
+
+class ClassPathUtil {
+
+    private static final int DEFAULT_CLASS_PATH_JAR_THRESHOLD = 1024 * 2;
+
+    private static final String ATTRIBUTE_MANIFEST_VERSION = "1.0";
+
+    private static final String MANIFEST_CLASS_PATH_SEPARATOR = " ";
+
+    private ClassPathUtil() {
+    }
+
+    static String retrieveCurrentClassPath(String workingFolderPath) throws IOException {
+
+        int classPathJarThreshold = findClassPathJarThreshold();
+
+        String fullClassPath = System.getProperty("java.class.path", "");
+
+        if (fullClassPath.length() < classPathJarThreshold) {
+            return fullClassPath;
+        } else {
+            return createClassPathJarFileAndReturnPath(fullClassPath, workingFolderPath);
+        }
+    }
+
+    private static int findClassPathJarThreshold() {
+        String classPathJarThresholdAsString = System.getProperty("quickperf.classPathJarThreshold");
+        if (classPathJarThresholdAsString != null) {
+            return Integer.parseInt(classPathJarThresholdAsString);
+        }
+        return DEFAULT_CLASS_PATH_JAR_THRESHOLD;
+    }
+
+    public static String createClassPathJarFileAndReturnPath(String fullClassPath, String workingFolderPath) throws IOException {
+
+        File classPathJar = new File(workingFolderPath + File.separator + "class_path.jar");
+
+        Manifest manifest = buildManifestFrom(fullClassPath);
+
+        createJarFrom(classPathJar, manifest);
+
+        return classPathJar.getAbsolutePath();
+
+    }
+
+    private static Manifest buildManifestFrom(String fullClassPath) throws MalformedURLException {
+        String classPath = formatClassPathForManifest(fullClassPath);
+        Manifest manifest = new Manifest();
+        Attributes mainAttributes = manifest.getMainAttributes();
+        mainAttributes.put(Attributes.Name.MANIFEST_VERSION, ATTRIBUTE_MANIFEST_VERSION);
+        mainAttributes.put(Attributes.Name.CLASS_PATH, classPath);
+        return manifest;
+    }
+
+    private static String formatClassPathForManifest(String fullClassPath) throws MalformedURLException {
+        String[] paths = fullClassPath.split(File.pathSeparator);
+        String classPath = "";
+        for (String path : paths) {
+            if (classPath.length() > 0) {
+                classPath += MANIFEST_CLASS_PATH_SEPARATOR;
+            }
+            classPath += new File(path).toURI().toURL().toString();
+        }
+        return classPath;
+    }
+
+    private static void createJarFrom(File classPathJar, Manifest manifest) throws IOException {
+        try (
+                FileOutputStream fos = new FileOutputStream(classPathJar);
+                BufferedOutputStream bos = new BufferedOutputStream(fos);
+                JarOutputStream jar = new JarOutputStream(bos, manifest);
+        ) {
+            //inspection IO resource opened are safely closed
+            jar.flush();
+        } catch (IOException io) {
+            throw io;
+        }
+    }
+}

--- a/core/src/main/java/org/quickperf/testlauncher/NewJvmTestLauncher.java
+++ b/core/src/main/java/org/quickperf/testlauncher/NewJvmTestLauncher.java
@@ -66,19 +66,23 @@ public class NewJvmTestLauncher {
 
         List<String> jvmOptionsAsStrings = jvmOptions.asStrings(workingFolder);
 
-        List<String> jvmCommand = buildCommand( mainClassArguments
-                                              , jvmOptionsAsStrings
-                                              , workingFolder.getPath()
-                                              , mainClassToLaunchTestInANewJvm);
-
-        return execute(jvmCommand);
+        try {
+            List<String> jvmCommand = buildCommand( mainClassArguments
+                                                  , jvmOptionsAsStrings
+                                                  , workingFolder.getPath()
+                                                  , mainClassToLaunchTestInANewJvm);
+            return execute(jvmCommand);
+        } catch (IOException e) {
+            return JvmIssue.buildFrom(e);
+        }
 
     }
 
+    //Large class path (more then 2K): fixed in ClassPathUtil using class path jar with manifest
     private List<String> buildCommand(MainClassArguments mainClassArguments
                                     , List<String> jvmOptionsAsStrings
                                     , String workingFolderPath
-                                    , Class<?> mainClassToLaunchTest) {
+                                    , Class<?> mainClassToLaunchTest) throws IOException {
         List<String> command = new ArrayList<>();
         command.add(retrieveJavaExePath());
         command.addAll(jvmOptionsAsStrings);
@@ -89,7 +93,7 @@ public class NewJvmTestLauncher {
                                     .buildForJvm(workingFolderPath)
                    );
         command.add("-cp");
-        command.add(retrieveCurrentClassPath());
+        command.add(ClassPathUtil.retrieveCurrentClassPath(workingFolderPath));
         command.add(mainClassToLaunchTest.getCanonicalName());
         List<String> mainClassArgumentsAsStringList = mainClassArguments.buildMainClassArgumentsForJvmCommand();
         command.addAll(mainClassArgumentsAsStringList);
@@ -101,10 +105,6 @@ public class NewJvmTestLauncher {
         return    javaHomeDirectoryPath
                 + File.separator + "bin"
                 + File.separator + "java";
-    }
-
-    private String retrieveCurrentClassPath() {
-        return System.getProperty("java.class.path");
     }
 
     private JvmIssue execute(List<String> cmd) {

--- a/junit5/junit5-jvm-test/src/test/java/org/quickperf/junit5/jvm/issue/ClassPathJarThresholdTest.java
+++ b/junit5/junit5-jvm-test/src/test/java/org/quickperf/junit5/jvm/issue/ClassPathJarThresholdTest.java
@@ -1,0 +1,50 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * Copyright 2019-2022 the original author or authors.
+ */
+package org.quickperf.junit5.jvm.issue;
+
+import org.junit.jupiter.api.Test;
+import org.quickperf.junit5.JUnit5Tests;
+import org.quickperf.junit5.JUnit5Tests.JUnit5TestsResult;
+import org.quickperf.junit5.QuickPerfTest;
+import org.quickperf.jvm.annotations.JvmOptions;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ClassPathJarThresholdTest {
+
+    @QuickPerfTest
+    static class ClassUnderTest {
+
+        @JvmOptions(value = "-Dquickperf.classPathJarThreshold=1")
+        @Test
+        public void a_test() {
+        }
+
+    }
+
+    @Test public void
+    should_not_fail_when_using_jar_class_path() {
+
+        // GIVEN
+        Class<?> testClass = ClassUnderTest.class;
+        JUnit5Tests jUnit5Tests = JUnit5Tests.createInstance(testClass);
+
+        // WHEN
+        JUnit5TestsResult jUnit5TestsResult = jUnit5Tests.run();
+
+        // THEN
+        assertThat(jUnit5TestsResult.getNumberOfFailures()).isZero();
+
+    }
+
+}

--- a/pom.xml
+++ b/pom.xml
@@ -416,6 +416,10 @@
             <name>Hedley Proctor</name>
             <url>https://github.com/hedleyproctor</url>
         </contributor>
+        <contributor>
+            <name>rsassi</name>
+            <url>https://github.com/rsassi</url>
+        </contributor>
     </contributors>
 
 </project>

